src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,337 @@
+    current_index         : Rc<Cell<usize>>
+}
+
+impl BreadcrumbsModel {
+    /// Create new BreadcrumbsModel.
+    pub fn new<'t,S:Into<&'t Scene>>(scene:S, frp:&Frp, focus_manager:&FocusManager) -> Self {
+        let scene                 = scene.into();
+        let project_name          = ProjectName::new(scene,focus_manager);
+        let logger                = Logger::new("Breadcrumbs");
+        let display_object        = display::object::Instance::new(&logger);
+        let breadcrumbs_container = display::object::Instance::new(&logger);
+        let animations            = Animations::new(&frp.network);
+        let scene                 = scene.clone_ref();
+        let breadcrumbs           = default();
+        let breadcrumb_pop        = frp.outputs.breadcrumb_pop.clone_ref();
+        let breadcrumb_push       = frp.outputs.breadcrumb_push.clone_ref();
+        let current_index         = default();
+        Self{logger,display_object,animations,scene,breadcrumbs,breadcrumb_pop,breadcrumb_push,
+            project_name,breadcrumbs_container,current_index}.init()
+    }
+
+    fn init(self) -> Self {
+        self.add_child(&self.project_name);
+        self.add_child(&self.breadcrumbs_container);
+        self.project_name.set_position(Vector3::new(HORIZONTAL_MARGIN,0.0,0.0));
+        self.set_project_name_width(self.project_name.width());
+        //FIXME[dg]: This is a workaround. Main shouldn't be a MethodPointer.
+        let mut method_pointer = enso_protocol::language_server::MethodPointer::default();
+        method_pointer.name    = "Main".to_string();
+        let method_pointer     = Rc::new(method_pointer);
+        self.push_breadcrumb(&method_pointer,&default());
+        self
+    }
+
+    fn width(&self) -> f32 {
+        self.breadcrumbs.borrow().iter().fold(0.0, |acc,breadcrumb| acc + breadcrumb.width())
+    }
+
+    fn set_project_name_width(&self, width:f32) {
+        self.breadcrumbs_container.set_position(Vector3::new(HORIZONTAL_MARGIN + width,0.0,0.0));
+    }
+
+    fn select_breadcrumb(&self, index:usize) {
+        let current_index = self.current_index.get();
+        match index.cmp(&current_index) {
+            cmp::Ordering::Less => {
+                // If we have more crumbs after `index`, we will pop them.
+                let pop_amount    = current_index - index;
+                for _ in 0..pop_amount {
+                    self.breadcrumb_pop.emit(());
+                }
+            },
+            cmp::Ordering::Greater => {
+                for index in current_index..index {
+                    let info = self.breadcrumbs.borrow().get(index).map(|breadcrumb| {
+                        (breadcrumb.info.method_pointer.clone(),breadcrumb.info.expression_id)
+                    }).as_ref().cloned();
+                    if let Some((method_pointer,expression_id)) = info {
+                        self.breadcrumb_push.emit((method_pointer,expression_id));
+                    }
```

Comment:
I don't think this is correct. If for some reason we cannot retrieve information for any crumb, we should stop at this, rather than skipping it and pushing remaining ones on top. This can only result in totally broken call stacks.
(though I know it shouldn't happen unless theres internal error)

Reply:
Yes, it should be fine, as long as the whole breadcrumbs panel stays in well-defined, well-behaved state after this.
Pushing only as many crumbs as we can is fine. (Even better would be removing the now-invalid ones from the "inactive" side and logging error.)